packages/labs/task/src/task.ts

```diff
@@ -400,16 +423,93 @@ export class Task<
         throw new Error(`Unexpected status: ${this.status}`);
     }
   }
-
-  private _argsDirty(args?: T) {
-    const prev = this._previousArgs;
-    this._previousArgs = args;
-    return Array.isArray(args) && Array.isArray(prev)
-      ? args.length === prev.length && args.some((v, i) => notEqual(v, prev[i]))
-      : args !== prev;
-  }
 }
 
 type MaybeReturnType<F> = F extends (...args: unknown[]) => infer R
   ? R
   : undefined;
+
+export const shallowArrayEquals = <T extends ReadonlyArray<unknown>>(
+  oldArgs: T,
+  newArgs: T
+) =>
+  oldArgs === newArgs ||
+  (oldArgs.length === newArgs.length &&
+    oldArgs.every((v, i) => !notEqual(v, newArgs[i])));
+
+export const deepArrayEquals = <T extends ReadonlyArray<unknown>>(
+  oldArgs: T,
+  newArgs: T
+) =>
+  oldArgs === newArgs ||
+  (oldArgs.length === newArgs.length &&
+    oldArgs.every((v, i) => deepEquals(v, newArgs[i])));
+
+const objectValueOf = Object.prototype.valueOf;
+const objectToString = Object.prototype.toString;
+const {keys: objectKeys} = Object;
+const {isArray} = Array;
+
+export const deepEquals = (a: unknown, b: unknown): boolean => {
```

Comment:
Deep equality checks require a number of opinionated pieces. Do we want to maintain this? Do we want to ship these bytes, or are we assuming production builds are tree shaked?

Reply:
Yes they do, but... it's quite hard to find a minimally good deep equals, especially that's shipped as JS modules. I think including some basic one is just good DX.

I do need to document the opinions here:
- Object.is for primitives
- Constructor check
- Arrays
- `valueOf()` and `toString()` checks
- Map and Set (whichever way we go)
- Own property requirement for objects
- Cycles (currently we don't detect them and just iloop, which a lot of libraries do, but some don't)

Reply:
WDYT about putting this in its own module?